# POPULAR RESTAURANTS IN JAIPUR

## AND ANALYSIS OF WHERE A NEWCOMER CAN OPEN HIS/HER RESTAURANTS NEAR POPULAR ONES

#### NOTE: THIS NOTEBOOK IS PART OF THE IBM CAPSTONE PROJECT: THE BATTLE OF NEIGHBOURHOODS

In [ ]:
# IMPORTING REQUIRED LIBRARIES

import numpy as np 

import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json 

!conda install -c conda-forge geopy --yes # this line installs geopy library for accessing longitude and latitude of a location
from geopy.geocoders import Nominatim 

import requests 
from pandas.io.json import json_normalize


import matplotlib.cm as cm
import matplotlib.colors as colors


#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library


In [2]:
#TESTING THE GEOPY LIBRARY AND GETTING THE LATITUDE AND LONGITUDE OF JAIPUR
address = 'Jaipur, Rajasthan, India'

geolocator = Nominatim(user_agent="jpr_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Jaipur is  {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Jaipur is  26.916194, 75.820349.


In [3]:
# INSTALLING THE BEAUTIFUL SOUP LIBRARY FOR SCRAPING THE WEB
conda install -c anaconda beautifulsoup4


Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.11

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - beautifulsoup4


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    soupsieve-1.9.2            |           py36_0          61 KB  anaconda
    openssl-1.0.2s             |       h7b6447c_0         3.1 MB  anaconda
    certifi-2019.6.16          |           py36_1         156 KB  anaconda
    beautifulsoup4-4.7.1       |           py36_1         143 KB  anaconda
    ------------------------------------------------------------
                                           Total:         3.5 MB

The following NEW packages will be INSTALLED:

    soupsieve:      1.9.2-py36_0      a

In [90]:
#IMPORTING THE LIBRARY
from bs4 import BeautifulSoup
import requests

In [91]:
#THIS IS THE URL THAT IS SCRAPED TO GET THE PINCODES AND DIFFERENT AREAS IN/NEAR JAIPUR
#IF THIS DOES NOT WORK IN THE FUTURE OR THE WEBSITE IS DOWN, I HAVE EXPORTED THE DATA TO A CSV FILE AND THAT TOO CAM BE USED
url='https://finkode.com/rj/jaipur.html'

In [92]:
page=requests.get(url) #get the url data

In [93]:
soup=BeautifulSoup(page.content,'html.parser') #the beautiul soup library's method is used to parse through the contents

In [94]:
#LISTS OF LOCATION AND PINCODE, THE LOOP BELOW APPENDS THE VALUE TO IT
Location=[]
Pincode=[]  

In [95]:
for row in soup.find('table').find_all('tr'):
    data = row.find_all('td')
    if(len(data) > 0):
        loc=data[0].text
        loc1=' '.join(loc.split())[:-3].upper()
        Location.append(loc1)
        Pincode.append(data[2].text)

#### Exporting the appended data to a data frame

In [96]:
dataF=pd.DataFrame({"Pincode":Pincode,"Location":Location})


In [98]:
#Updating the first location name to Jobner 
dataF.at[0,'Location']= "Jobner"

#First Five Locations and their pincode
dataF.head()


Pincode     Location
0  303329       Jobner
1  302006  AJMER ROAD 
2  302013    AKHEPURA 
3  303338       AKODA 
4  303901     AKODIYA

In [99]:
# THESE COMMENTED LINES OF CODE USE THE GEOPY LIBRARY TO EXTRACT THE LONGITUDE AND LATITUDE OF EACH LOCATION FROM THE SCRAPED DATA
#abs                            in THE dataF DATAFRAME
# lat=[]
# long=[]
# locat=[]
# for loc in dataF['Location']:
#     address=loc+", Jaipur, Rajasthan, India"
#     geolocator = Nominatim(user_agent="jp_explorer")
#     location = geolocator.geocode(address)
#     if location==None:
#         continue
#     else:
#        lat.append(location.latitude)
#        long.append(location.longitude)
#        locat.append(loc)
       
     
# dataF2=pd.DataFrame({"Location":locat,"Latitude":lat,"Longitude":long})

# SINCE THAT HAS BEEN ALREADY EXPORTED TO A CSV FILE I PREFER USING THAT AS THE LOADING HAPPENS QUICKLY
dataF2=pd.read_csv("export_dataframe.csv")



In [101]:
print(dataF2) # THE LONGITUDE AND LATITUDE OF DIFFERENT LOCATIONS in JAIPUR

                         Location   Latitude  Longitude
0                          Jobner  26.973031  75.386994
1                     AJMER ROAD   26.916532  75.799666
2                      AMBA BARI   26.941420  75.775581
3                      AMER ROAD   27.833340  76.257263
4                           AMER   26.988827  75.855892
5                          ANDHI   27.058825  76.163387
6                        ASALPUR   26.906004  75.498493
7                         BADHAL   27.251411  75.450871
8                 BAGAWAS AHIRAN   27.533493  76.073095
9                        BAGAWAS   27.533493  76.073095
10                         BAGRU   26.808795  75.542312
11                   BAJAJ NAGAR   26.875099  75.809849
12                         BANAR   26.620495  75.063240
13                     BANI PARK   26.931655  75.795801
14                    BAPU NAGAR   26.885073  75.800958
15                  BARKAT NAGAR   26.878536  75.785600
16                       BHAWANI   26.893557  75

In [19]:
#export_csv = dataF2.to_csv (r'export_dataframe.csv', index = None, header=True) #Don't forget to add '.csv' at the end of the path
# IGNORE IF EXPORTED

In [102]:
dataF4=pd.merge(dataF2,dataF,on='Location') #MERGING THE dataF data frame with the dataF2 on location

In [103]:
dataF4

Location   Latitude  Longitude Pincode
0                          Jobner  26.973031  75.386994  303329
1                     AJMER ROAD   26.916532  75.799666  302006
2                      AMBA BARI   26.941420  75.775581  302039
3                      AMER ROAD   27.833340  76.257263  302002
4                           AMER   26.988827  75.855892  302028
5                          ANDHI   27.058825  76.163387  303001
6                        ASALPUR   26.906004  75.498493  303604
7                         BADHAL   27.251411  75.450871  303602
8                 BAGAWAS AHIRAN   27.533493  76.073095  303119
9                        BAGAWAS   27.533493  76.073095  303603
10                         BAGRU   26.808795  75.542312  303007
11                   BAJAJ NAGAR   26.875099  75.809849  302015
12                         BANAR   26.620495  75.063240  303105
13                     BANI PARK   26.931655  75.795801  302016
14                    BAPU NAGAR   26.885073  75.800958  302015
15                  BARKAT NAGAR   26.878536  75.785600  302015
16                       BHAWANI   26.893557  75.811360  303001
17                         BOBAS   26.913598  75.567674  303338
18                        CHAKSU   26.605983  75.943034  303901
19                         CHOMU   27.172500  75.721500  303702
20                         DHAND   26.950888  75.324216  302028
21                         DHULA   26.919293  75.824036  303305
22                          DUDU   26.681122  75.232076  303008
23                     DURGAPURA   26.850627  75.784527  302018
24                          GARH   27.240297  75.645913  303302
25                     GHASIPURA   26.717513  76.087309  303103
26                         GONER   26.852823  75.897172  303905
27                     GOPALPURA   26.875264  75.143609  303006
28                     GOPALPURA   26.875264  75.143609  303108
29                     GOPALPURA   26.875264  75.143609  303006
30                     GOPALPURA   26.875264  75.143609  303108
31                        HATHOJ   26.959304  75.679610  302012
32                      HINGONIA   27.012111  75.493014  303328
33                       HIRNODA   26.886013  75.327286  303338
34                     JAGATPURA   26.833578  75.827304  303806
35                     JAGATPURA   26.833578  75.827304  302017
36                     JAGATPURA   26.833578  75.827304  303806
37                     JAGATPURA   26.833578  75.827304  302017
38                        JAHOTA   27.122138  75.714729  303701
39                   JAIPUR CITY   26.916194  75.820349  302003
40                     JAIPUR G.P  26.916194  75.820349  302001
41                   JAIPUR R.S.   26.916194  75.820349  302006
42                  JAISINGHPURA   26.951629  75.890150  303801
43                  JAISINGHPURA   26.951629  75.890150  303108
44                  JAISINGHPURA   26.951629  75.890150  303106
45                  JAISINGHPURA   26.951629  75.890150  303801
46                  JAISINGHPURA   26.951629  75.890150  303108
47                  JAISINGHPURA   26.951629  75.890150  303106
48                  JAISINGHPURA   26.951629  75.890150  303801
49                  JAISINGHPURA   26.951629  75.890150  303108
50                  JAISINGHPURA   26.951629  75.890150  303106
51                  JAITPURA I A   27.122138  75.714729  303704
52                         JALSU   27.113858  75.628717  303701
53                  JANTA COLONY   26.916194  75.820349  302004
54                       JATWARA   26.869395  76.217242  303305
55                 JAWAHAR NAGAR   26.893043  75.836196  302004
56      JHOTWARA INDUSTRIAL AREA   26.958586  75.752769  302012
57                      JHOTWARA   26.958586  75.752769  302012
58                        JOBNER   26.973031  75.386994  303328
59                      KACHRODA   26.876761  75.263939  303338
60                    KALYANPURA   26.819961  75.772309  303108
61                    KALYANPURA   26.819961  75.772309  

In [104]:
df_duplicates = pd.DataFrame.drop_duplicates(dataF4) #REMOVING THE DUPLICATES

In [18]:
df_duplicates

Location   Latitude  Longitude Pincode
0                          Jobner  26.973031  75.386994  303329
1                     AJMER ROAD   26.916532  75.799666  302006
2                      AMBA BARI   26.941420  75.775581  302039
3                      AMER ROAD   27.833340  76.257263  302002
4                           AMER   26.988827  75.855892  302028
5                          ANDHI   27.058825  76.163387  303001
6                        ASALPUR   26.906004  75.498493  303604
7                         BADHAL   27.251411  75.450871  303602
8                 BAGAWAS AHIRAN   27.533493  76.073095  303119
9                        BAGAWAS   27.533493  76.073095  303603
10                         BAGRU   26.808795  75.542312  303007
11                   BAJAJ NAGAR   26.875099  75.809849  302015
12                         BANAR   26.620495  75.063240  303105
13                     BANI PARK   26.931655  75.795801  302016
14                    BAPU NAGAR   26.885073  75.800958  302015
15                  BARKAT NAGAR   26.878536  75.785600  302015
16                       BHAWANI   26.893557  75.811360  303001
17                         BOBAS   26.913598  75.567674  303338
18                        CHAKSU   26.605983  75.943034  303901
19                         CHOMU   27.172500  75.721500  303702
20                         DHAND   26.950888  75.324216  302028
21                         DHULA   26.919293  75.824036  303305
22                          DUDU   26.681122  75.232076  303008
23                     DURGAPURA   26.850627  75.784527  302018
24                          GARH   27.240297  75.645913  303302
25                     GHASIPURA   26.717513  76.087309  303103
26                         GONER   26.852823  75.897172  303905
27                     GOPALPURA   26.875264  75.143609  303006
28                     GOPALPURA   26.875264  75.143609  303108
31                        HATHOJ   26.959304  75.679610  302012
32                      HINGONIA   27.012111  75.493014  303328
33                       HIRNODA   26.886013  75.327286  303338
34                     JAGATPURA   26.833578  75.827304  303806
35                     JAGATPURA   26.833578  75.827304  302017
38                        JAHOTA   27.122138  75.714729  303701
39                   JAIPUR CITY   26.916194  75.820349  302003
40                     JAIPUR G.P  26.916194  75.820349  302001
41                   JAIPUR R.S.   26.916194  75.820349  302006
42                  JAISINGHPURA   26.951629  75.890150  303801
43                  JAISINGHPURA   26.951629  75.890150  303108
44                  JAISINGHPURA   26.951629  75.890150  303106
51                  JAITPURA I A   27.122138  75.714729  303704
52                         JALSU   27.113858  75.628717  303701
53                  JANTA COLONY   26.916194  75.820349  302004
54                       JATWARA   26.869395  76.217242  303305
55                 JAWAHAR NAGAR   26.893043  75.836196  302004
56      JHOTWARA INDUSTRIAL AREA   26.958586  75.752769  302012
57                      JHOTWARA   26.958586  75.752769  302012
58                        JOBNER   26.973031  75.386994  303328
59                      KACHRODA   26.876761  75.263939  303338
60                    KALYANPURA   26.819961  75.772309  303108
61                    KALYANPURA   26.819961  75.772309  303503
64                        KANOTA   26.912313  75.929688  303012
65                        KANSLI   27.688852  76.136490  303108
66                      KARANSAR   27.094257  75.449258  303603
67                       KHANDEL   26.952660  75.279575  303604
68                     KHATIPURA   26.836121  75.891071  303301
69                     KHATIPURA   26.836121  75.891071  302012
72                KHATIPURA ROAD   26.933797  75.753617  302006
73                      KHEJROLI   27.338733  75.680178  303803
74                         KHORI   26.851379  75.908514  303103
75             KISHANGARH RENWAL   27.159715  75.359364  

### GROUPING THE LOCATIONS WITH SAME PINCODE

In [105]:
dataNeighbour = df_duplicates.groupby(["Pincode", "Longitude","Latitude"], as_index=False).agg(lambda x: ", ".join(x))
dataNeighbour.head(20)
#Grouping Neighbours with one PinCode

Pincode  Longitude   Latitude                              Location
0   302001  75.774971  26.978949                RAJASTHAN STATE HOTEL 
1   302001  75.820349  26.916194                            JAIPUR G.P
2   302002  76.257263  27.833340                            AMER ROAD 
3   302003  75.820349  26.916194                          JAIPUR CITY 
4   302004  75.820349  26.916194                         JANTA COLONY 
5   302004  75.836196  26.893043                        JAWAHAR NAGAR 
6   302006  75.753617  26.933797                       KHATIPURA ROAD 
7   302006  75.799666  26.916532                           AJMER ROAD 
8   302006  75.820349  26.916194                          JAIPUR R.S. 
9   302012  75.679610  26.959304                               HATHOJ 
10  302012  75.734870  26.920460                                SIRSI 
11  302012  75.752769  26.958586  JHOTWARA INDUSTRIAL AREA , JHOTWARA 
12  302012  75.891071  26.836121                            KHATIPURA 
13  302013  75.784055  26.992197          VISHWAKARMA INDUSTRIAL AREA 
14  302015  75.785600  26.878536                         BARKAT NAGAR 
15  302015  75.800958  26.885073                           BAPU NAGAR 
16  302015  75.809849  26.875099                          BAJAJ NAGAR 
17  302016  75.795801  26.931655                            BANI PARK 
18  302016  75.818192  26.893944                        SINDHI COLONY 
19  302016  75.820349  26.916194                        SHASTRI NAGAR

## MAP OF JAIPUR

In [21]:
# create map of Jaipur using latitude and longitude values

address = 'Jaipur, Rajasthan, India'

geolocator = Nominatim(user_agent="jpr_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Jaipur is  {}, {}.'.format(latitude, longitude))



map_jaipur = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, loc, in zip(dataNeighbour['Latitude'], dataNeighbour['Longitude'], dataNeighbour['Location']):
    label = '{}'.format(loc)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_jaipur)  
    
map_jaipur 

The geograpical coordinate of Jaipur is  26.916194, 75.820349.


### MY CLIENT ID AND CLIENT SECRET
#### YOU WON'T BE ABLE TO USE THESE AS I WILL RESET THEM TO NEWER ONES
##### YOU CAN CREATE YOUR OWN FOURSQUARE API, JUST GOOGLE IT

In [106]:

CLIENT_ID = 'HXRNOSOKCO5PNHYVW05HJQZP0YSWH5VCSRORZHGDSLCQO31C' # your Foursquare ID
CLIENT_SECRET = 'V1S14PGWFL3RCGPPSTNPGXJDEXJEZS3CJGA4UXAN3IE0HEPA' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: HXRNOSOKCO5PNHYVW05HJQZP0YSWH5VCSRORZHGDSLCQO31C
CLIENT_SECRET:V1S14PGWFL3RCGPPSTNPGXJDEXJEZS3CJGA4UXAN3IE0HEPA


## FOOD VENUES IN JAIPUR

In [107]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [108]:
#Food venues in Jaipur
#GENERATING THE URL TO GET THE FOOD VENUES IN JAIPUR

neighbourhood_latitude=26.916194
neighbourhood_longitude=75.820349
LIMIT = 200 # limit of number of venues returned by Foursquare API
radius = 15000 # define radius
section='food'
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&section={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius,
    section,
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=HXRNOSOKCO5PNHYVW05HJQZP0YSWH5VCSRORZHGDSLCQO31C&client_secret=V1S14PGWFL3RCGPPSTNPGXJDEXJEZS3CJGA4UXAN3IE0HEPA&v=20180605&ll=26.916194,75.820349&radius=15000&section=food&limit=200'

### GETTING THE JSON RESULT OF THE URL

In [109]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d4d0fe8f8953d0026ce6dbe'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Jaipur',
  'headerFullLocation': 'Jaipur',
  'headerLocationGranularity': 'city',
  'query': 'food',
  'totalResults': 55,
  'suggestedBounds': {'ne': {'lat': 27.051194135000134,
    'lng': 75.97146783539611},
   'sw': {'lat': 26.781193864999867, 'lng': 75.66923016460387}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '5623621e498e8ec2b86fc810',
       'name': 'Curious Life Coffee Roasters',
       'location': {'address': 'P 25, Yudhisthir Marg,',
        'crossStreet': 'C Scheme',
        'lat': 26.904097278407598,
        'lng': 75.79712119777699,
        'labeledLat

## FILTERING OUT THE DATA AND EXTRACTING THE NAME , CATEGORY AND LATITUDE AND LONGITUDE OF THE RESAUTRANTS

In [110]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

name                     categories  \
0         Curious Life Coffee Roasters                           Café   
1         NIBS- The Chocolate Workshop                           Café   
2                      Peacock Rooftop              Indian Restaurant   
3                          Brown Sugar                         Bakery   
4                          Anokhi Cafe                           Café   
5                          Spice Court              Indian Restaurant   
6                             Doorbeen             Italian Restaurant   
7                        Old Take Away                      BBQ Joint   
8                             Peshawri              Indian Restaurant   
9                Rawat Mishtan Bhandar                    Snack Place   
10                           Mr. Beans                           Café   
11                     Cafe Coffee Day                           Café   
12            Egg Dee - Sanjay Omlette              Indian Restaurant   
13                                 OTH                         Bakery   
14                            Cinnamon              Indian Restaurant   
15                          Brewberrys                           Café   
16                            Taruveda                           Café   
17                           C'est bon                           Café   
18                             Saffron              Indian Restaurant   
19              Lakshmi Mithai Bhandar              Indian Restaurant   
20           Kalyan Rooftop Restaurant              Indian Restaurant   
21                      Caffe Palladio             Turkish Restaurant   
22       Nibs - The Chocolate Workshop                         Bakery   
23                    Kanha restaurant              Indian Restaurant   
24                        Little italy             Italian Restaurant   
25                              Chaisa                           Café   
26                         Brown Sugar                         Bakery   
27                               Kanha              Indian Restaurant   
28  Surabhi Restaurant & Turban Museum              Indian Restaurant   
29          The Forresta Kitchen & Bar                     Restaurant   
30                     Barbeque Nation                      BBQ Joint   
31                               Niros              Indian Restaurant   
32              Surya Mahal (Bake Hut)              Indian Restaurant   
33                     The Marble Arch                     Restaurant   
34                    Handi Restaurant              Indian Restaurant   
35                     Mamu's Infusion             Italian Restaurant   
36                        Four Seasons  Vegetarian / Vegan Restaurant   
37                           Pizza Hut                    Pizza Place   
38                       Cafe LazyMojo                           Café   
39                           Pizza Hut                    Pizza Place   
40                 Indian Coffee House                           Café   
41                       Saras Parlour                    Snack Place   
42                                 F32                     Restaurant   
43                           Dana Pani  Vegetarian / Vegan Restaurant   
44                    Kanha Restaurant           Fast Food Restaurant   
45                      Domino's Pizza                    Pizza Place   
46                             1135 AD              Indian Restaurant   
47                     cafe coffee day                           Café   
48                     Cafe Coffee Day                           Café   
49                          Airr Toast                  Deli / Bodega   
50                  Palki Poolside Bar                     Restaurant   
51                 Pratap Plaza,Jaipur                     Food Court   
52                                Zoya                     Restaurant   
53               Madeira Bar & Terrace                     Restaurant   
54                 

## MAP OF RESTAURANTS IN JAIPUR

In [78]:
# create map restraunt in Jaipur
address = 'Jaipur, Rajasthan, India'

geolocator = Nominatim(user_agent="jpr_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Jaipur is  {}, {}.'.format(latitude, longitude))



map_jaipur = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, in zip(nearby_venues['lat'], nearby_venues['lng']):
    label = '{}'.format(loc)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='purple',
        fill=True,
        fill_color='white',
        fill_opacity=0.7,
        parse_html=False).add_to(map_jaipur)  
    
map_jaipur

The geograpical coordinate of Jaipur is  26.916194, 75.820349.


### LETS SELECT A PARTICULAR PLACE IN JAIPUR AND SEARCH FOR TOP VENUES WITHIN 3000m OF THAT PLACE

In [112]:
#Top venues within radius of 3000m in Jhotwara

neighbourhood_latitude=26.958586
neighbourhood_longitude=75.752769
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 3000 # define radius
section='food'
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&section={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    section,
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=HXRNOSOKCO5PNHYVW05HJQZP0YSWH5VCSRORZHGDSLCQO31C&client_secret=V1S14PGWFL3RCGPPSTNPGXJDEXJEZS3CJGA4UXAN3IE0HEPA&v=20180605&ll=26.958586,75.752769&radius=3000&section=food&limit=100'

In [113]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d4d112c6e46500033d20906'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Jaipur',
  'headerFullLocation': 'Jaipur',
  'headerLocationGranularity': 'city',
  'query': 'food',
  'totalResults': 10,
  'suggestedBounds': {'ne': {'lat': 26.98558602700003,
    'lng': 75.78300413244109},
   'sw': {'lat': 26.931585972999972, 'lng': 75.72253386755891}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '53876471498e26479aba1334',
       'name': "Domino's Pizza",
       'location': {'address': 'Shop No.G-11,12,13,14,87,88,89,90,Ground Floor, Alankar Plaza,A-10,Central Spine, Vidhyadhar Nagar ,Jaipur- 302016(Rajasthan)',
        'lat': 26.957,
        'lng

In [65]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

name           categories        lat  \
0                       Domino's Pizza          Pizza Place  26.957000   
1                            Pizza Hut          Pizza Place  26.963242   
2                      Cafe Coffee Day                 Café  26.962145   
3                            Foodcourt           Food Court  26.956493   
4           Kanji Hotel And Restaurant    Indian Restaurant  26.963415   
5  Sunflower Restaurant Hotel Paradise  American Restaurant  26.947934   
6                      Spice and curry     Asian Restaurant  26.947953   
7                       Chowdhry Hotel          Snack Place  26.972876   
8                     Kaku Sa's Palace    Indian Restaurant  26.934877   

         lng  
0  75.778000  
1  75.779319  
2  75.779295  
3  75.771905  
4  75.771530  
5  75.773608  
6  75.773642  
7  75.772291  
8  75.749280

### MAP OF NEARBY VENUES IN JHOTWARA

In [115]:
address = 'Jhotwara,Jaipur, Rajasthan, India'

geolocator = Nominatim(user_agent="jpr_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Jaipur is  {}, {}.'.format(latitude, longitude))



map_jaipur = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, in zip(nearby_venues['lat'], nearby_venues['lng']):
    label = '{}'.format(loc)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='orange',
        fill=True,
        fill_color='green',
        fill_opacity=0.7,
        parse_html=False).add_to(map_jaipur)  
    
map_jaipur

The geograpical coordinate of Jaipur is  26.9585861, 75.7527694.


## KANHA IS A POPULAR RESTAURANT IN JAIPUR, LETS EXPLORE 

In [116]:
# Exploring popular venues around Kanha Restaurant
latitude=26.903506
longitude=75.753229

In [117]:
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=HXRNOSOKCO5PNHYVW05HJQZP0YSWH5VCSRORZHGDSLCQO31C&client_secret=V1S14PGWFL3RCGPPSTNPGXJDEXJEZS3CJGA4UXAN3IE0HEPA&ll=26.903506,75.753229&v=20180605&radius=3000&limit=100'

In [118]:
results = requests.get(url).json()
'There are {} around Kanha restaurant.'.format(len(results['response']['groups'][0]['items']))

'There are 21 around Kanha restaurant.'

In [119]:
items = results['response']['groups'][0]['items']
items[0]

{'reasons': {'count': 0,
  'items': [{'summary': 'This spot is popular',
    'type': 'general',
    'reasonName': 'globalInteractionReason'}]},
 'venue': {'id': '4c0220c6cbd876b086f0f043',
  'name': 'Kanha restaurant',
  'location': {'address': 'Queens Road.',
   'lat': 26.903505994979938,
   'lng': 75.75322941910966,
   'labeledLatLngs': [{'label': 'display',
     'lat': 26.903505994979938,
     'lng': 75.75322941910966}],
   'distance': 0,
   'cc': 'IN',
   'city': 'Jaipur',
   'state': 'Rājasthān',
   'country': 'India',
   'formattedAddress': ['Queens Road.', 'Jaipur', 'Rājasthān', 'India']},
  'categories': [{'id': '4bf58dd8d48988d10f941735',
    'name': 'Indian Restaurant',
    'pluralName': 'Indian Restaurants',
    'shortName': 'Indian',
    'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/indian_',
     'suffix': '.png'},
    'primary': True}],
  'photos': {'count': 0, 'groups': []}},
 'referralId': 'e-0-4c0220c6cbd876b086f0f043-0'}

In [120]:
dataframe = json_normalize(items) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# filter the category for each row
dataframe_filtered['venue.categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean columns
dataframe_filtered.columns = [col.split('.')[-1] for col in dataframe_filtered.columns]

dataframe_filtered.head(10)

name                     categories  \
0    Kanha restaurant              Indian Restaurant   
1      Baskin Robbins                 Ice Cream Shop   
2           Dana Pani  Vegetarian / Vegan Restaurant   
3         burger farm           Fast Food Restaurant   
4  INOX Elements Mall                      Multiplex   
5              Subway                 Sandwich Place   
6               rj 14        North Indian Restaurant   
7           Pizza Hut                    Pizza Place   
8     Cafe Coffee Day                    Coffee Shop   
9         Ibis Jaipur                          Motel   

                                            address        lat        lng  \
0                                      Queens Road.  26.903506  75.753229   
1                                    Vaishali Nagar  26.897825  75.751119   
2                             Janpath , Shyam Nagar  26.891594  75.759807   
3                                      Queens House  26.905036  75.753015   
4                        Third Floor, Elements Mall  26.894975  75.748900   
5                                   Evershine Tower  26.911496  75.743335   
6                                        Ajmer road  26.895414  75.750355   
7  Shop No 20, Evershine Tower, F1, Amrapali Circle  26.911537  75.743296   
8                                     Vaibhav tower  26.912121  75.739059   
9                                       Civil Lines  26.908228  75.779899   

                                      labeledLatLngs  distance  cc    city  \
0  [{'label': 'display', 'lat': 26.90350599497993...         0  IN  Jaipur   
1  [{'label': 'display', 'lat': 26.89782528967085...       666  IN     NaN   
2  [{'label': 'display', 'lat': 26.89159418519215...      1478  IN     NaN   
3  [{'label': 'display', 'lat': 26.905036, 'lng':...       171  IN     NaN   
4  [{'label': 'display', 'lat': 26.89497482319445...      1042  IN  Jaipur   
5  [{'label': 'display', 'lat': 26.91149556954461...      1324  IN  Jaipur   
6  [{'label': 'display', 'lat': 26.895414, 'lng':...       944  IN  Jaipur   
7  [{'label': 'display', 'lat': 26.91153729254961...      1331  IN  Jaipur   
8  [{'label': 'display', 'lat': 26.91212075597084...      1702  IN  Jaipur   
9  [{'label': 'display', 'lat': 26.90822817191843...      2699  IN  Jaipur   

       state country                                   formattedAddress  \
0  Rājasthān   India           [Queens Road., Jaipur, Rājasthān, India]   
1        NaN   India                            [Vaishali Nagar, India]   
2        NaN   India                     [Janpath , Shyam Nagar, India]   
3        NaN   India                              [Queens House, India]   
4  Rājasthān   India  [Third Floor, Elements Mall (Ajmer Road), Jaip...   
5  Rājasthān   India  [Evershine Tower (Vaishali), Jaipur 302021, Rā...   
6  Rājasthān   India             [Ajmer road, Jaipur, Rājasthān, India]   
7  Rājasthān   India  [Shop No 20, Evershine Tower, F1, Amrapali Cir...   
8  Rājasthān   India  [Vaibhav tower (Vaishali nagar), Jaipur, Rājas...   
9  Rājasthān   India  [Civil Lines (Ajmer Road), Jaipur, Rājasthān, ...   

      crossStreet postalCode                        id  
0             NaN        NaN  4c0220c6cbd876b086f0f043  
1             NaN        NaN  4ecbce6093adb4bd1b52c582  
2             NaN        NaN  4ecf685bd3e3521ce4703398  
3             NaN        NaN  576f8c96498e84c74e80850c  
4      Ajmer Road     302018  52c6e60f498ee459a4da729e  
5        Vaishali     302021  5118dd85e4b0d2e90bcbd871  
6             NaN        NaN  588b65ad1d21ba2202524414  
7  Vaishali Nagar     302021  50e82191e4b0b2f92a886377  
8  Vaishali nagar        NaN  4c706a54f52cb713f19fafaf  
9      Ajmer Road        NaN  5253cf8711d28a6b882aea47

## MAP OF TOP RESTAURANTS AROUND KANHA

#### THE BLACK DOT IS KANHA ON THE MAP

In [125]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=15) # generate map centred around Ecco


# add Kanha as a violet circle mark
folium.features.CircleMarker(
    [latitude, longitude],
    radius=10,
    popup='KANHA',
    fill=True,
    color='black',
    fill_color='black',
    fill_opacity=0.6
    ).add_to(venues_map)


# add popular spots to the map as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        fill=True,
        color='red',
        fill_color='blue',
        fill_opacity=0.6
        ).add_to(venues_map)

# display map
venues_map

# OBSERVATIONS:

AS WE CAN SEE FROM THE MAP AND THE FILTERED DATA FRAME ABOVE, EVEN THE NEAREST TOP RESTAURANTS ARE 1000 meters far on average.

THE ONLY VENUE THAT IS WITHIN 1 KM is BASKIN AND ROBINS ICE CREAM SHOP AT 666 meter from KANHA.

SINCE JAIPUR IS A GROWING CITY AND PEOPLE FROM OUTER STATES ARE MIGRATING AND INVESTING RAPIDLY, 
IT IS A GOLDEN PERIOD FOR PEOPLE INTERESTED IN OPENING A FOOD BUSINESS TO LOOK AROUND THE CITY AND GO AHEAD 